activerecord/lib/active_record/relation.rb

```diff
@@ -582,6 +582,11 @@ def ==(other)
         records == other
       end
     end
+    alias eql? ==
+
+    def hash
+      records.hash
```

Comment:
This seems rather unlike the `==` method above.

I'm also not sure we want `eql?` to be quite so generous; `Relation#==(Array)` is convenient, but I think `eql?` should only be true for something more clearly Relationy.

Reply:
I think you are correct. I did an alias since that seemed to be the common behavior in other parts of Rails but looking against, I don't think the `==` method is quite the right implementation. We probably only want it to return true if they are both relation objects.

Also we want it looser than `==` in some ways. The `==` methods only consider them `==` if their SQL is the same if they are both relation objects. But I think we want to allow them to possibly have different SQL that arrives at the same records and still consider them `eql?`. This would make them behave more like `Array#eql?`.

If it seems reasonable to you guys I would like to just move this entirely to the [`activerecord/relation/delegation.rb` file](https://github.com/rails/rails/blob/3bcdd09ed47778b928d6d51e621e33d75682827d/activerecord/lib/active_record/relation/delegation.rb#L41-L44) and just add both `eql?` and `hash` to the list of methods delegated.

This will give it the exact same semantics as `Array` which is what I think we want.

Reply:
What if we moved it down a level to an `AssociationRelation`? This is really my use case as I am grouping records by their associated objects. I.E.

```
class Widget < ApplicationRecord
  has_many :properties
end

class Property < ApplicationRecord
  belong_to :widget
end
```

```
Widget.all.group_by &:properties
```

I.E. I want to group all records that have the same properties.

The semantics between `Relation#==` and `AssociationRelation#==` are different. In `Relation#==` it doesn't materialize the records while in `AssociationRelation#==` it does. I think it is surprising that they have different semantics (and it leads to the complicated `==` implementation in `Relation` as it has to deal with comparing a `Relation` and a `AssociationRelation`). But that is besides the point.

Since `AssociationRelation#==` does load the records it would also seem reasonable that `AssociationRelation#eql?` would load the records. And that would at least allow that type of relation to be grouped (my use case and probably the most common use case). This way we limit the scope of the PR but also hopefully don't surprise someone else down the road when they try to group on associated records like the existing implementation surprised me.